src/memory.rs

```diff
@@ -0,0 +1,153 @@
+//! Utilities for converting Kubernetes quantities to Java heap settings.
+
+use k8s_openapi::apimachinery::pkg::api::resource::Quantity;
+
+use crate::error::{Error, OperatorResult};
+use std::str::FromStr;
+
+#[derive(Clone, Debug, PartialEq, PartialOrd)]
+pub enum BinaryMultiple {
+    Kibi,
+    Mebi,
+    Gibi,
+    Tebi,
+    Pebi,
+    Exbi,
+}
+
+impl BinaryMultiple {
+    pub fn to_legacy(&self) -> String {
+        match self {
+            BinaryMultiple::Kibi => "k".to_string(),
+            BinaryMultiple::Mebi => "m".to_string(),
+            BinaryMultiple::Gibi => "g".to_string(),
+            BinaryMultiple::Tebi => "t".to_string(),
+            BinaryMultiple::Pebi => "p".to_string(),
+            BinaryMultiple::Exbi => "e".to_string(),
+        }
+    }
+
+    pub fn upscale(&self) -> Self {
+        match self {
+            BinaryMultiple::Kibi => BinaryMultiple::Kibi,
+            BinaryMultiple::Mebi => BinaryMultiple::Kibi,
+            BinaryMultiple::Gibi => BinaryMultiple::Mebi,
+            BinaryMultiple::Tebi => BinaryMultiple::Gibi,
+            BinaryMultiple::Pebi => BinaryMultiple::Tebi,
+            BinaryMultiple::Exbi => BinaryMultiple::Pebi,
+        }
+    }
+}
+
+impl FromStr for BinaryMultiple {
+    type Err = Error;
+
+    fn from_str(q: &str) -> OperatorResult<BinaryMultiple> {
+        let lq = q.to_lowercase();
+        match lq.as_str() {
+            "ki" | "kib" => Ok(BinaryMultiple::Kibi),
+            "mi" | "mib" => Ok(BinaryMultiple::Mebi),
+            "gi" | "gib" => Ok(BinaryMultiple::Gibi),
+            "ti" | "tib" => Ok(BinaryMultiple::Tebi),
+            "pi" | "pib" => Ok(BinaryMultiple::Pebi),
+            "ei" | "eib" => Ok(BinaryMultiple::Exbi),
+            _ => Err(Error::InvalidQuantityUnit {
+                value: q.to_string(),
+            }),
+        }
+    }
+}
+
+/// Easily transform K8S memory resources to Java heap options.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Memory {
+    value: f32,
+    unit: BinaryMultiple,
+}
+
+impl Memory {
+    /// Scale by the given factor. If the factor is less then one
+    /// the unit granularity is increased one level to ensure eventual
+    /// conversions to Java heap settings don't end up with zero values..
```

Comment:
This scaling seems like something we'd want to do when rounding/formatting the value for output?